Book additional files support

werwolfby opened the pull request:
## Additional Files Support (1 of 3 PRs)

This pull request introduces support for managing additional files (such as alternative formats and supplementary materials) associated with books in the BookLore API. The changes include new entities, DTOs, repository methods, service logic, and controller endpoints to enable uploading, listing, downloading, and deleting these files. The Book model and mapping logic have also been updated to expose these files via the API.

This is a continuation of the work for #221 and serves as the first step that officially reserves a migration number for this functionality.

### Considerations and Architecture

To add support for additional files to existing books, I considered two options:

#### Option 1: Extract file data from the book itself

The idea was to extract all columns in the `book` table related to file information into a separate table. Currently, the `book` table contains the following columns related to file information: `file_name`, `file_sub_path`, `file_size_kb`, `initial_hash`, and `current_hash`. These could be moved to a separate table with a reference back to the `book` entity.

**Pros**:

* Allows storing as many book files as needed.
* All files and their hashes are stored in a single table with a unique index on `current_hash`.

**Cons**:

* Introduces significant complexity and requires refactoring of existing code.
* Makes managing the main book file more complex, as consistency is harder to maintain (e.g., ensuring at least one file exists for each book).
* Adds an extra join to the files table for most queries involving books.
  * Since most users do not require additional files per book, this could negatively impact performance for the majority of users without providing any benefit.

#### Option 2: Add additional files as a separate entity

This option involves creating a new entity for additional files, allowing the existing book structure to remain intact while providing a flexible way to manage multiple files associated with each book.

**Pros**:

* Simpler implementation with no need to refactor the existing book structure.
* Clear separation of concerns, enabling easy management of additional files without affecting the main book file.
* Allows adding, removing, or modifying additional files without impacting the core book entity.
* No extra joins for most book queries, as the main book file remains unchanged.
* Keeps performance optimal for most users, as additional files are only queried when explicitly requested.

**Cons**:

* Requires two tables for book files with `current_hash`, which could lead to some data duplication.

#### Decision

After evaluating both options, I decided to proceed with **Option 2**. This approach introduces additional files without complicating the existing book structure while providing the flexibility needed for future enhancements. It also ensures that the majority of users will not experience any performance degradation, as additional files are only queried when explicitly requested.

## Technical Details

**Additional File Management Functionality**

* Added the `BookAdditionalFileEntity` entity to represent additional files linked to books, with fields for file metadata and a method to construct the full file path. (`booklore-api/src/main/java/com/adityachandel/booklore/model/entity/BookAdditionalFileEntity.java`)
* Introduced the `AdditionalFile` DTO and the `AdditionalFileType` enum to model and classify additional files as either alternative formats or supplementary files. (`booklore-api/src/main/java/com/adityachandel/booklore/model/dto/AdditionalFile.java`, `booklore-api/src/main/java/com/adityachandel/booklore/model/enums/AdditionalFileType.java`) [[1]](diffhunk://#diff-7252a62174581ad5901d577a6be377330b003d8ab8ef1dc5c14645e3dfd04d55R1-R23) [[2]](diffhunk://#diff-9c2b24db83f54f98a94e784543aa907080be3e5ce9c5b71ca37c40dc78cfd022R1-R5)
* Added the `BookAdditionalFileRepository` to provide queries for retrieving, counting, and searching additional files by various criteria. (`booklore-api/src/main/java/com/adityachandel/booklore/repository/BookAdditionalFileRepository.java`)

**API and Service Layer Enhancements**

* Implemented the `AdditionalFileService` to handle listing, downloading, and deleting additional files, including physical file operations and database updates. (`booklore-api/src/main/java/com/adityachandel/booklore/service/AdditionalFileService.java`)
* Added the `AdditionalFileController` to expose REST endpoints for managing additional files: listing by book and type, uploading (via multipart/form-data), downloading, and deleting files, with appropriate access checks. (`booklore-api/src/main/java/com/adityachandel/booklore/controller/AdditionalFileController.java`)

**Model and Mapping Updates**

* Updated the `BookEntity` and `Book` DTO to include lists of associated additional files, split into `alternativeFormats` and `supplementaryFiles`. (`booklore-api/src/main/java/com/adityachandel/booklore/model/entity/BookEntity.java`, `booklore-api/src/main/java/com/adityachandel/booklore/model/dto/Book.java`) [[1]](diffhunk://#diff-2bee240ccf86912e043e6f4b69b30865aca660cf786414e8fa010089ffb04026R82-R84) [[2]](diffhunk://#diff-b9aa706556651103666200cc1cba69c7e6ba6de7ed7c15630da2c6232437392eR37-R38)
* Enhanced the `BookMapper` and introduced `AdditionalFileMapper` to map between entities and DTOs, with logic to filter and transform additional files by type. (`booklore-api/src/main/java/com/adityachandel/booklore/mapper/BookMapper.java`, `booklore-api/src/main/java/com/adityachandel/booklore/mapper/AdditionalFileMapper.java`) [[1]](diffhunk://#diff-c4d4ad9235f7608600c23d88ca14234b7349bd066b8af05e6f3e61040b0f6e33R3-R45) [[2]](diffhunk://#diff-c4d4ad9235f7608600c23d88ca14234b7349bd066b8af05e6f3e61040b0f6e33R69-R101) [[3]](diffhunk://#diff-e3d4aaf22d6f02aad0f7188ba42da67541ac53da6aa5659d6d24c8d829f19402R1-R30)

**Testing and Dependency Updates**

* Added Testcontainers dependencies for MariaDB and JUnit Jupiter to support integration testing of file-related features. (`booklore-api/build.gradle`)

These changes collectively provide a robust backend foundation for handling additional files associated with books, including secure access, file storage, and API exposure.

## Special Notes

### Test Containers

This PR introduces Testcontainers for MariaDB to facilitate integration testing of the additional files functionality. The `BookAdditionalFileRepositoryTest` class uses a MariaDB container to run tests against a real database instance, ensuring that the repository methods work correctly with the actual database schema and data.

### `current_hash` Uniqueness

The `current_hash` field in the `BookAdditionalFileEntity` is designed to be unique across additional formats only. Supplementary files can have the same hash.

Modern databases like PostgreSQL support partial indexes, but MariaDB does not. As a result, some workarounds were applied to ensure that `current_hash` is unique only for additional formats.

A new calculated column was added:

```sql
alt_format_current_hash     VARCHAR(128) AS (CASE WHEN additional_file_type = 'ALTERNATIVE_FORMAT' THEN current_hash END) STORED
```

And a unique index was created on this column:

```sql
UNIQUE INDEX idx_book_additional_file_current_hash_alt_format (alt_format_current_hash)
```

Unfortunately, MariaDB cannot optimize queries for `current_hash` with this index, so we have to use `alt_format_current_hash` in queries.

> [!NOTE]
>
> I use [https://cloudnative-pg.io/](CloudNativePG) for a home lab PostgreSQL instance. It provides features like automatic failover, backups, and more. It is an excellent tool for running PostgreSQL in production or development environments.
> 
> I have not found a similar tool for MariaDB that offers the same level of features and ease of use. While MariaDB is a great database, it lacks some advanced features available in PostgreSQL.
> 
> I hope BookLore will migrate to PostgreSQL in the future, as it offers better support for advanced features like partial indexes, JSONB, and more. Additionally, it is more widely adopted as far as I can tell.
> 
> **I can assist with the migration process, including automating data migration from MariaDB to PostgreSQL for our users.**

adityachandelgit commented:
Hello, I took a look at your code and the idea looks great! I haven’t had a chance to run the code or test the UI yet. I do have a few questions though:
	1.	How do features like Bookdrop, file move, metadata refresh, etc., work with your changes?
	2.	Have you ensured that all existing app features continue working correctly and nothing is broken?

Yes, I am definitely planning to move to PostgreSQL. Initially, I thought it was too heavy for a simple Booklore, but now that Booklore has become more complex, I’d definitely want to migrate to it. Doing this in a backwards compatible way will be a bit challenging though.

werwolfby commented:
@adityachandelgit thanks for the feedback — let me clarify the approach.

One of the main reasons I split this into 3 separate PRs and chose the second approach was specifically to keep the existing implementation untouched and introduce only new functionality.

> How do features like Bookdrop, file move, metadata refresh, etc., work with your changes?

* **Bookdrop** – supported for `Book` only. I don’t currently see a viable way to add additional files via Bookdrop.
* **File move** – I overlooked this and will check it separately.
* **Metadata refresh** – not affected, thanks to option 2.

Because `Book` remains a separate entity and additional files are optional, everything should behave exactly as before for existing `Book` functionality. Possible corner cases (like file move) can be handled in the second PR (#870). This PR’s focus is mainly on reserving the migration version and adding the API for additional files.

> Have you ensured that all existing app features continue working correctly and nothing is broken?

As noted in the 3rd PR (#872), the new feature should be considered beta and will require more testing — it’s a large change. To reduce the risk, I created a separate processor for the “Folder as Book” library type, where one file is the main book and others are additional files. This type is only used if the user explicitly selects it, so the default behavior remains unchanged.

werwolfby commented:
@adityachandelgit any news?

adityachandelgit commented:
@werwolfby 

Hey, sorry for the delay, things have been pretty busy on my end so I haven’t had much time to focus on Booklore. I’ll run your changes locally in the next couple of days and make sure everything works as expected with both the existing features and the new additions. Really appreciate all the effort you’ve put into this!

adityachandelgit commented:
@werwolfby 

I’m running into a few errors with PRs 2 and 3, and the Flyway migration file versions also need to be updated.
To avoid conflicts, I won’t be adding new Flyway files on my side until your PRs are merged, so it would be great if we can get these finalized soon.

```
✘ [ERROR] NG3004: Unable to import component Header.
  The symbol is not exported from /Users/aditya.chandel/Workspace/booklore/booklore-ui/node_modules/primeng/api/index.d.ts (module 'primeng/api'). [plugin angular-compiler]

    src/app/book/components/additional-file-uploader/additional-file-uploader.component.ts:47:13:
      47 │ export class AdditionalFileUploaderComponent implements OnInit, On...
         ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  The component is declared here.

    ../../node_modules/primeng/api/shared.d.ts:4:0:
      4 │ export declare class Header {
        ╵ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

For quicker back-and-forth, feel free to hop onto the official Booklore Discord: https://discord.gg/Ee5hd458Uz

werwolfby commented:
I've rebased the PRs, and fixed the migration version. I've pointed specifically:

> This is a continuation of the work for https://github.com/booklore-app/booklore/issues/221 and serves as the first step that officially reserves a migration number for this functionality.

To avoid such issues this PR first of all reserve the migration number :) Update version is easy, so I'm doing it many times already for this PRs as well.

Yes, can confirm `npm` build issue, working on should be fixed soon. Didn't update the PR after migrate to new `primeng` after #903.

werwolfby commented:
@adityachandelgit I fixed the `npm` build issue as well (it belongs to the second PR #870).

Will run some additional testing later

werwolfby commented:
@adityachandelgit rebased and verified all new functionality working as expected.
